### PR TITLE
Chapter 3 exercises Oscar KAN-12

### PR DESCRIPTION
**Pull request by insert-oreyes**
- chapter3 setup: update of the db and new chapter-3 folder
Screenshot 2023-11-01 at 11 38 59

- chapter exercises: all 3 working and test passed

Screenshot 2023-11-01 at 11 39 06

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (jportella93) [inline .vscode/settings.json]: `.vscode` folder shouldn't be commited, since thisare user-specific preferences and not project-specific. More info https://blog.martinhujer.cz/dont-put-idea-vscode-directories-to-projects-gitignore/

File: constants/chapter.json [hunk 2]
```diff
@@ -47,9 +47,43 @@
...
     "subtitle": "Functions",
     "image": "/chapter-3.jpg",
-    "exercises": "",
+    "exercises": "Minimum, Recursion, Bean counting",
     "link": "/chapter3",
-    "details": ""
+    "details": "This chapter taught you how to write your own functions. Separating the tasks your program performs into different functions is helpful. You won’t have to repeat yourself as much, and functions can help organize a program by grouping code into pieces that do specific things.",
+    "devOscar": [
+      {
+        "id": 1,
```
Comment (jportella93) [inline constants/chapter.json:55]: Feels like we don't need these `id`s since the title is already unique and can be used as an id?

File: constants/chapter.json [hunk 3]
```diff
@@ -47,9 +47,43 @@
...
+        "title": "Minimum",
+        "details": "Press the button below to display the youngest person."
+      },
+      {
+        "id": 2,
+        "title": "Recursion",
+        "details": "So I was thinking about recursion, then I found a flower and I came up with this:"
```
Comment (jportella93) [inline constants/chapter.json]: For consistency, I would remove the trailing `:`

File: .vscode/settings.json (hunk 1 above)
Comment (insert-oreyes) [reply .vscode/settings.json]: Deleted

File: pages/chapter3.tsx [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+const chapterMatch = chapter.find((chapter) => chapter.id === chapterId);
+
+if (!chapterMatch) {
+  throw new Error(`Chapter ${chapterId} not found`);
+}
+
+const { title, subtitle, details, devOscar, devJon } = chapterMatch;
+
+export default function Chapter3() {
```
Comment (jportella93) [inline pages/chapter3.tsx:20]: Can we add a test for Chapter3? it should go into `pageTests/Chapter3.test.tsx`

File: constants/chapter.json (hunk 2 above)
Comment (insert-oreyes) [reply constants/chapter.json:55]: I am using this id to link with every page, if I remove it, it will crash

File: components/exercises/chapter-3-oscar/Recursion.tsx [hunk 5]
```diff
@@ -0,0 +1,47 @@
+import React, { useEffect, useState } from 'react';
+import './Recursion.css';
+
+export function isEven(param: number): boolean {
```
Comment (jportella93) [inline components/exercises/chapter-3-oscar/Recursion.tsx:8]: Can we add docs for this function? You should be able to do it quickly with github copilot `/docs` command

File: components/exercises/chapter-3-oscar/Recursion.tsx [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  } else if (param < 0) {
+    return isEven(-param);
+  } else {
+    return isEven(param - 2);
+  }
+}
+
+const petals: JSX.Element[] = [];
+
+const Recursion = () => {
```
Comment (jportella93) [inline components/exercises/chapter-3-oscar/Recursion.tsx]: Can we add docs?

Comment (jportella93) [inline components/exercises/chapter-3-oscar/Recursion.tsx]: Also, can we use `export default function Recursion() {`

File: components/exercises/chapter-3-oscar/Recursion.tsx [hunk 7]
```diff
@@ -0,0 +1,47 @@
+import React, { useEffect, useState } from 'react';
+import './Recursion.css';
```
Comment (jportella93) [inline components/exercises/chapter-3-oscar/Recursion.tsx]: Do we need this CSS? can't we use Tailwind?

File: components/exercises/chapter-3-oscar/Recursion.tsx [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    if (isEven(timeRemaining) === false) {
+      petals.push(<div className='petal bg-[#b2990b9f]' />);
+    } else {
+      petals.push(<div className='petal bg-[#b2430b9f]' />);
+    }
+  };
+
+  return (
+    <div className='wrap'>
+      {petals.map((petal) => petal)}
```
Comment (jportella93) [inline components/exercises/chapter-3-oscar/Recursion.tsx:66]: This doesn't do anything, might as well just do 
```
{petals}
```

File: components/exercises/chapter-3-oscar/Recursion.tsx [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    const timer = setTimeout(() => {
+      petalAdd();
+      setTimeRemaining(
+        timeRemaining === 8 ? timeRemaining + 0 : timeRemaining + 1
+      );
+    }, 1000);
+    return () => clearTimeout(timer);
+  });
+
+  const petalAdd = () => {
```
Comment (jportella93) [inline components/exercises/chapter-3-oscar/Recursion.tsx]: better name `addPetal`

File: components/exercises/chapter-3-oscar/Recursion.tsx [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+      petals.push(<div className='petal bg-[#b2990b9f]' />);
+    } else {
+      petals.push(<div className='petal bg-[#b2430b9f]' />);
+    }
+  };
+
+  return (
+    <div className='wrap'>
+      {petals.map((petal) => petal)}
+      <div className='center'></div>
```
Comment (jportella93) [inline components/exercises/chapter-3-oscar/Recursion.tsx]: Does this do anything?

File: components/exercises/chapter-3-oscar/Recursion.test.tsx [hunk 11]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+import { render, screen } from '@testing-library/react';
+import Recursion, { isEven } from './Recursion';
+
+describe('isEven', () => {
```
Comment (jportella93) [inline components/exercises/chapter-3-oscar/Recursion.test.tsx:3]: Can we add a test for negative numbers?

File: components/exercises/chapter-3-oscar/Recursion.tsx (hunk 8 above)
Comment (insert-oreyes) [reply components/exercises/chapter-3-oscar/Recursion.tsx:66]: it doesn't work, that is why I mapped it

File: components/exercises/chapter-3-oscar/BeanCounting.tsx [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+  let count = 0;
+  for (let i = 0; i < param.length; i++) {
+    if (param[i] === char) {
+      count++;
+    }
+  }
+  return count;
+}
+
+const BeanCounting = () => {
```
Comment (jportella93) [inline components/exercises/chapter-3-oscar/BeanCounting.tsx]: export default function

File: components/exercises/chapter-3-oscar/BeanCounting.tsx [hunk 13]
```diff
@@ -0,0 +1,47 @@
+import React, { useState } from 'react';
+
+export function countBs(param: string): number {
```
Comment (jportella93) [inline components/exercises/chapter-3-oscar/BeanCounting.tsx:12]: Add docs for each function, you can use github copilot /doc

File: components/exercises/chapter-3-oscar/Recursion.tsx (hunk 9 above)
Comment (insert-oreyes) [reply components/exercises/chapter-3-oscar/Recursion.tsx]: refactored

File: components/exercises/chapter-3-oscar/BeanCounting.test.tsx [hunk 14]
```diff
@@ -0,0 +1,17 @@
+import { countBs, countChar } from './BeanCounting';
```
Comment (jportella93) [inline components/exercises/chapter-3-oscar/BeanCounting.test.tsx]: No test for `BeanCounting`?

File: components/exercises/chapter-3-oscar/Minimum.tsx [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+  },
+  {
+    name: 'Jacob',
+    age: 23,
+    image: '/avatar2.png',
+    phrase: 'Hi there! I am Jacob, I am 23 years old.',
+  },
+];
+
+const getMinimum = (a: number, b: number) => Math.min(a, b);
```
Comment (jportella93) [inline components/exercises/chapter-3-oscar/Minimum.tsx]: I think the point of the exercise is to create a fn that doesn't use `Math.min`

>The [previous chapter](https://eloquentjavascript.net/02_program_structure.html#return_values) introduced the standard function Math.min that returns its smallest argument. We can build something like that now. Write a function min that takes two arguments and returns their minimum.

File: components/exercises/chapter-3-oscar/Minimum.tsx [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+
+const stateFiltered = initialState.filter((item) =>
+  item.age === getMinimum(initialState[0].age, initialState[1].age)
+    ? item
+    : null
+);
+
+const Minimum = () => {
```
Comment (jportella93) [inline components/exercises/chapter-3-oscar/Minimum.tsx]: export default function

File: components/exercises/chapter-3-oscar/Minimum.tsx [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+  const [state, setState] = useState(initialState);
+
+  return (
+    <div>
```
Comment (jportella93) [inline components/exercises/chapter-3-oscar/Minimum.tsx]: do we need this extra empty div?

File: components/exercises/chapter-3-oscar/Minimum.tsx [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+const getMinimum = (a: number, b: number) => Math.min(a, b);
+
+const stateFiltered = initialState.filter((item) =>
+  item.age === getMinimum(initialState[0].age, initialState[1].age)
+    ? item
+    : null
+);
+
+const Minimum = () => {
+  const [state, setState] = useState(initialState);
```
Comment (jportella93) [inline components/exercises/chapter-3-oscar/Minimum.tsx]: Better naming 
```js
const [people, setPeople] = useState(initialPeople);
```

File: components/exercises/chapter-3-oscar/Minimum.tsx [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+
+  return (
+    <div>
+      <div className='wrapper flex flex-col'>
+        <div className='flex flex-row justify-between'>
+          {state.map((item) => (
```
Comment (jportella93) [inline components/exercises/chapter-3-oscar/Minimum.tsx]:
```js
{people.map((person) => (
```

File: constants/chapter.json (hunk 3 above)
Comment (insert-oreyes) [reply constants/chapter.json]: refactored

File: components/exercises/chapter-3-oscar/Minimum.test.tsx [hunk 20]
```diff
@@ -0,0 +1,30 @@
+import { render, screen } from '@testing-library/react';
+import Minimum from './Minimum';
+
+describe('Minimum component', () => {
+  it('renders the name of the youngest person', () => {
```
Comment (jportella93) [inline components/exercises/chapter-3-oscar/Minimum.test.tsx:5]: Incorrect. Since we haven't pressed the button yet, both people are shown.

Can we update these tests so they assert:
- Initially, both people are shown
- When the button is pressed, only the youngest is shown

**Review by jportella93 — COMMENTED** [review]
Added some comments

File: components/exercises/chapter-3-oscar/Recursion.tsx (hunk 8 above)
Comment (jportella93) [reply components/exercises/chapter-3-oscar/Recursion.tsx:66]: That might be revealing a bug, maybe related to having the `petals` const outside the component. Can we update it so we don't have to use `{petals.map((petal) => petal)}`? That line is a code smell

File: pages/chapter3.tsx (hunk 4 above)
Comment (insert-oreyes) [reply pages/chapter3.tsx:20]: Test added

File: components/exercises/chapter-3-oscar/Recursion.tsx (hunk 5 above)
Comment (insert-oreyes) [reply components/exercises/chapter-3-oscar/Recursion.tsx:8]: Docs added

File: components/exercises/chapter-3-oscar/Recursion.tsx (hunk 6 above)
Comment (insert-oreyes) [reply components/exercises/chapter-3-oscar/Recursion.tsx]: docs added

Comment (insert-oreyes) [reply components/exercises/chapter-3-oscar/Recursion.tsx]: Refactored

File: components/exercises/chapter-3-oscar/Recursion.tsx (hunk 7 above)
Comment (insert-oreyes) [reply components/exercises/chapter-3-oscar/Recursion.tsx]: Refactored

File: components/exercises/chapter-3-oscar/Recursion.tsx (hunk 8 above)
Comment (insert-oreyes) [reply components/exercises/chapter-3-oscar/Recursion.tsx:66]: I tried, but nothing happened, could I have any debugging advice? Thank you

File: components/exercises/chapter-3-oscar/Recursion.tsx (hunk 10 above)
Comment (insert-oreyes) [reply components/exercises/chapter-3-oscar/Recursion.tsx]: That is the middle of the flower

File: components/exercises/chapter-3-oscar/Recursion.test.tsx (hunk 11 above)
Comment (insert-oreyes) [reply components/exercises/chapter-3-oscar/Recursion.test.tsx:3]: added and passed

File: components/exercises/chapter-3-oscar/BeanCounting.tsx (hunk 12 above)
Comment (insert-oreyes) [reply components/exercises/chapter-3-oscar/BeanCounting.tsx]: Refactored

File: components/exercises/chapter-3-oscar/BeanCounting.tsx (hunk 13 above)
Comment (insert-oreyes) [reply components/exercises/chapter-3-oscar/BeanCounting.tsx:12]: docs added

File: components/exercises/chapter-3-oscar/BeanCounting.test.tsx (hunk 14 above)
Comment (insert-oreyes) [reply components/exercises/chapter-3-oscar/BeanCounting.test.tsx]: test added and passed

File: components/exercises/chapter-3-oscar/Minimum.tsx (hunk 15 above)
Comment (insert-oreyes) [reply components/exercises/chapter-3-oscar/Minimum.tsx]: refactored

File: components/exercises/chapter-3-oscar/Minimum.tsx (hunk 16 above)
Comment (insert-oreyes) [reply components/exercises/chapter-3-oscar/Minimum.tsx]: refactored

File: components/exercises/chapter-3-oscar/Minimum.tsx (hunk 17 above)
Comment (insert-oreyes) [reply components/exercises/chapter-3-oscar/Minimum.tsx]: you are right, refactored

File: components/exercises/chapter-3-oscar/Minimum.tsx (hunk 18 above)
Comment (insert-oreyes) [reply components/exercises/chapter-3-oscar/Minimum.tsx]: refactored

File: components/exercises/chapter-3-oscar/Minimum.tsx (hunk 19 above)
Comment (insert-oreyes) [reply components/exercises/chapter-3-oscar/Minimum.tsx]: refactored

File: components/exercises/chapter-3-oscar/Minimum.test.tsx (hunk 20 above)
Comment (insert-oreyes) [reply components/exercises/chapter-3-oscar/Minimum.test.tsx:5]: added

File: app/globals.css [hunk 21]
```diff
@@ -137,3 +137,30 @@
 .footer__copyrights-link {
   @apply flex-1 flex sm:justify-end justify-center max-sm:mt-4 gap-10;
 }
+/* Recursion css styles */
+.petal-rotating {
```
Comment (jportella93) [inline app/globals.css]: Can you move this in a CSS file next to the tested component `recursion.css` and then import it in `Recursion.tsx` by doing `import "./recursion.css"?

File: components/Footer.test.tsx [hunk 22]
```diff
@@ -11,7 +11,7 @@ describe('Footer', () => {
   it('renders the footer links', () => {
     render(<Footer />);
     const footerLinks = screen.getAllByRole('link');
-    expect(footerLinks).toHaveLength(18);
+    expect(footerLinks).toHaveLength(14);
```
Comment (jportella93) [inline components/Footer.test.tsx:14]: Why did this change? we didn't change anything in the footer

File: components/exercises/chapter-3-oscar/BeanCounting.test.tsx [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+  });
+});
+
+describe('BeanCounting', () => {
+  it('renders the phrase value 0, when the input does not include B', () => {
+    const phraseValue = 'Hello, world!';
+    render(<BeanCounting phraseValue={phraseValue} />);
+    expect(screen.getByText('This phrase has 0 B letter')).toBeInTheDocument();
+  });
+  it('renders the phrase value 1, when the input includes B', () => {
```
Comment (jportella93) [inline components/exercises/chapter-3-oscar/BeanCounting.test.tsx:27]: Can you add blank lines between tests for readibility?

File: components/exercises/chapter-3-oscar/BeanCounting.tsx [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+  }
+  return count;
+}
+
+/**
+ * Renders a component that takes a phrase as input and displays the number of 'B' and 's' letters in it.
+ * @returns JSX.Element
+ */
+
+export default function BeanCounting({ phraseValue }: BeanCountingProps) {
```
Comment (jportella93) [inline components/exercises/chapter-3-oscar/BeanCounting.tsx]: No docs for the phraseValue prop?

File: components/exercises/chapter-3-oscar/Minimum.tsx [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+  },
+  {
+    name: 'Jacob',
+    age: 23,
+    image: '/avatar2.png',
+    phrase: 'Hi there! I am Jacob, I am 23 years old.',
+  },
+];
+
+// const getMinimum = (a: number, b: number) => Math.min(a, b);
```
Comment (jportella93) [inline components/exercises/chapter-3-oscar/Minimum.tsx]: commented out code?

File: components/exercises/chapter-3-oscar/Minimum.tsx [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+const getMinimum = (a: number, b: number) => (a < b ? a : b);
+
+const stateFiltered = initialPeople.filter((item) =>
+  item.age === getMinimum(initialPeople[0].age, initialPeople[1].age)
+    ? item
+    : null
+);
+
+export default function Minimum() {
```
Comment (jportella93) [inline components/exercises/chapter-3-oscar/Minimum.tsx:44]: no docs?

File: components/exercises/chapter-3-oscar/Recursion.tsx (hunk 8 above)
Comment (jportella93) [reply components/exercises/chapter-3-oscar/Recursion.tsx:66]: That's okay let's leave it as is for now

File: constants/chapter.json (hunk 2 above)
Comment (jportella93) [reply constants/chapter.json:55]: Ok

File: components/Footer.test.tsx (hunk 22 above)
Comment (jportella93) [reply components/Footer.test.tsx:14]: Ah I see, the changes below. This should go on a separate PR

File: pageTests/Chapter3.test.tsx [hunk 27]
```diff
@@ -0,0 +1,33 @@
+import { render, screen } from '@testing-library/react';
+import Chapter3 from '../pages/chapter3';
```
Comment (jportella93) [inline pageTests/Chapter3.test.tsx:2]: can you add a newline?

**Review by jportella93 — APPROVED** [review]
Some comments, otherwise LGTM

File: app/globals.css (hunk 21 above)
Comment (insert-oreyes) [reply app/globals.css]: Moved, thank you

File: components/Footer.test.tsx (hunk 22 above)
Comment (insert-oreyes) [reply components/Footer.test.tsx:14]: I changed as test was complaining about the usage of the same word in different fields

File: components/exercises/chapter-3-oscar/BeanCounting.test.tsx (hunk 23 above)
Comment (insert-oreyes) [reply components/exercises/chapter-3-oscar/BeanCounting.test.tsx:27]: added

File: components/exercises/chapter-3-oscar/BeanCounting.tsx (hunk 24 above)
Comment (insert-oreyes) [reply components/exercises/chapter-3-oscar/BeanCounting.tsx]: docs included

File: components/exercises/chapter-3-oscar/Minimum.tsx (hunk 25 above)
Comment (insert-oreyes) [reply components/exercises/chapter-3-oscar/Minimum.tsx]: comment deleted

File: components/exercises/chapter-3-oscar/Minimum.tsx (hunk 26 above)
Comment (insert-oreyes) [reply components/exercises/chapter-3-oscar/Minimum.tsx:44]: docs added

File: pageTests/Chapter3.test.tsx (hunk 27 above)
Comment (insert-oreyes) [reply pageTests/Chapter3.test.tsx:2]: line added
